raft/src/main/java/org/apache/kafka/raft/internals/SerdeRecordsIterator.java

```diff
@@ -0,0 +1,231 @@
+    private final Records records;
+    private final RecordSerde<T> serde;
+    private final BufferSupplier bufferSupplier;
+    private final int maxBatchSize;
+
+    private Optional<Iterator<MutableRecordBatch>> nextBatches = Optional.empty();
+    private Optional<Batch<T>> nextBatch = Optional.empty();
+    // Buffer used to as the backing store for nextBatches if needed
+    private Optional<ByteBuffer> allocatedBuffer = Optional.empty();
+    // Number of bytes from records that read
+    private int bytesRead = 0;
+    private boolean isClosed = false;
+
+    public SerdeRecordsIterator(
+        Records records,
+        RecordSerde<T> serde,
+        BufferSupplier bufferSupplier,
+        int maxBatchSize
+    ) {
+        this.records = records;
+        this.serde = serde;
+        this.bufferSupplier = bufferSupplier;
+        this.maxBatchSize = maxBatchSize;
+    }
+
+    @Override
+    public boolean hasNext() {
+        checkIfClosed();
+
+        if (!nextBatch.isPresent()) {
+            nextBatch = nextBatch();
+        }
+
+        return nextBatch.isPresent();
+    }
+
+    @Override
+    public Batch<T> next() {
+        if (!hasNext()) {
+            throw new NoSuchElementException("Batch iterator doesn't have any more elements");
+        }
+
+        Batch<T> batch = nextBatch.get();
+        nextBatch = Optional.empty();
+
+        return batch;
+    }
+
+    @Override
+    public void close() {
+        isClosed = true;
+        allocatedBuffer.ifPresent(bufferSupplier::release);
+        allocatedBuffer = Optional.empty();
+    }
+
+    private void checkIfClosed() {
+        if (isClosed) {
+            throw new IllegalStateException("Serde record batch itererator was closed");
+        }
+    }
+
+    private Optional<Iterator<MutableRecordBatch>> nextBatches() {
+        int recordSize = records.sizeInBytes();
+        if (bytesRead < recordSize) {
+            final MemoryRecords memoryRecords;
+            if (records instanceof MemoryRecords) {
+                bytesRead = recordSize;
+                memoryRecords = (MemoryRecords) records;
+            } else if (records instanceof FileRecords) {
+                final ByteBuffer buffer;
+                if (allocatedBuffer.isPresent()) {
+                    buffer = allocatedBuffer.get();
+                    buffer.compact();
+
+                    if (!buffer.hasRemaining()) {
+                        // The buffer is not big enough to read an entire batch
+                        throw new IllegalStateException(
+                            String.format(
+                                "Unable to read batch from file records buffer %s with maximum batch %s and record size %s",
+                                buffer,
+                                maxBatchSize,
+                                records.sizeInBytes()
+                            )
+                        );
+                    }
```

Comment:
@hachikuji @mumrah I am not super happy with this handling. Specially since it is possible for us to write a batch greater than `maxBatchSize` based on how `BatchAccumulator` works. I am thinking for now that we configure the write size to 8MB and the read size to something like 10MB.

When we have a better solution for `scheduleAtomicAppend` that doesn't rely on batches for atomicity, I think this is less important.

Reply:
Can we just let the buffer grow as needed? We have a `BufferSupplier` after all.

Reply:
To get this new size we need to ready the size from the batch record using the `Readable` for example. I don't know if we want to trust this value for memory allocation.

Reply:
We have a history of size-related allocation failures when reading from the log, so it would be nice to avoid them. I agree we should probably not allocate blindly, but I also do not think we should put too much faith in the max batch size. Eventually the max batch size may be configurable, but we'll always have to deal with historic data.

Reply:
Yeah, got it. I thought about it some more. This is currently only used for the metadata partitions. If the controller or broker cannot read the metadata partition then it cannot make any progress in the rest of the system. In this case, the process should probably shutdown. Causing the process to shutdown because of an `OOMException` is probably okay.